### PR TITLE
Throw an error if next or push is called after nil

### PR DESCRIPTION
**Pull request by ibash**
This implements this comment: https://github.com/caolan/highland/issues/207#issuecomment-71706695

After `push(null _.nil)` is called for a generator, calling `push` or `next` throws an error.
Adding the test and implementation revealed another bug in `zipAll` that would try to write to the stream after nil has been pushed. That bug was fixed.

### CODE REVIEW

**Comment by ibash** [comment]
cc @vqvu / @quarterto

File: test/test.js [hunk 1]
```diff
@@ -470,6 +470,34 @@ exports['async next from consumer'] = function (test) {
     });
 };
 
+exports['generator throws error if next called after nil'] = function(test) {
+    var s = _(function(push, next) {
```
Comment (vqvu) [inline test/test.js]: As a general stylistic thing, can you put a `test.expect(1)` here? Same with the other test.

**Comment by vqvu** [comment]
Can you explain why the change to `zipAll` is needed? What's the test case? It just feels like you're fixing the wrong thing here.

**Comment by ibash** [comment]
It turns out that zipAll would call push / next after nil was pushed to the stream. This would happen because of the for loop at the bottom.

File: lib/index.js [hunk 2]
```diff
@@ -407,10 +407,22 @@ function Stream(/*optional*/xs, /*optional*/ee, /*optional*/mappingHint) {
     }
     else if (typeof xs === 'function') {
         this._generator = xs;
+        this._generator_nil_seen = false;
```
Comment (vqvu) [inline lib/index.js]: Also, I'd call this `_nil_seen`, since it's not a generator-specific state.

**Comment by vqvu** [comment]
Can you give a test case that provokes this behavior?

**Comment by vqvu** [comment]
Oh, I see. If two streams pushed `nil` at once, the `nil` will get pushed twice. But that has nothing to do with the loop at the bottom...

The fix for that would be

```javascript
else if (x === _.nil) {
    if (!finished) {
        finished = true;
        push(null, nil);
    }
}
```

**Comment by ibash** [comment]
Take a look at the other PR -- thoughts? I also have a couple edge cases for throwing an error if push/next is called after nil for consumers. However there are some problems with trying to write a fix for this: https://github.com/ibash/highland/commit/1a14d8d7182ef294ce25d66814c5f432100ec4ae

**Comment by vqvu** [comment]
Ok, I figured out why the various `latest` tests were failing in a different test now. It's because `test.done` was called too soon.

Example: In the `latest - let errors pass through` test.

```javascript
var errs = [];
var s2 = s.latest().errors(function (err) {
    errs.push(err);
});
var s3 = s2.consume(function (err, x, push, next) {
    push(err, x);
    setTimeout(next, 60); // This is bug
});
s3.toArray(function (xs) {
    // values at 0s, 60s, 120s
    test.same(xs, [1, 1, 'last']);
    test.same(errs, ['foo', 'bar']);

    // Done should not be called here.
    test.done();
});
this.clock.tick(1000);
// Done should be called here. Which will cause the error to show up under the correct test.
```

When you update this PR, can you also make this change? Thanks!

**Comment by vqvu** [comment]
Also, about your work on ibash@ef725fb117e416e39d239a410799bfee89d29df3.

You should just re-use `s._nil_seen` (the one used by the generator checks) instead of closing over a new `nil_seen` variable. It'll let other parts of the stream access that state if it needs to.

**Comment by ibash** [comment]
Roger that -- I'll bundle up the fixes for this and then consumer fixes into one PR

**Comment by ibash** [comment]
closing in favor of https://github.com/caolan/highland/pull/220
